nav2_util/include/nav2_util/simple_action_server.hpp

```diff
@@ -95,30 +102,60 @@ class SimpleActionServer
 
           if (is_active(pending_handle_)) {
             debug_msg("The pending slot is occupied."
-              " The previous pending goal will be aborted and replaced.");
-
-            pending_handle_->abort(empty_result());
-            pending_handle_.reset();
-            preempt_requested_ = false;
+              " The previous pending goal will be terminated and replaced.");
+            terminate(pending_handle_);
           }
-
-          debug_msg("Setting flag so the action server can grab the preempt request.");
-          preempt_requested_ = true;
           pending_handle_ = handle;
+          preempt_requested_ = true;
         } else {
           if (is_active(pending_handle_)) {
             // Shouldn't reach a state with a pending goal but no current one.
-            error_msg("Forgot to handle a preemption. Aborting the pending goal.");
-
-            pending_handle_->abort(empty_result());
-            pending_handle_.reset();
+            error_msg("Forgot to handle a preemption. Terminating the pending goal.");
+            terminate(pending_handle_);
             preempt_requested_ = false;
           }
 
-          debug_msg("Starting a thread to process the goals");
-
           current_handle_ = handle;
-          std::thread{execute_callback_}.detach();
+
+          auto work = [this]() {
```

Comment:
You're creating lambdas within lambdas here. It's time to make some actual functions.

Reply:
I think if you just turn these three or four lambdas in the constructor into methods, that would be good enough. 

If you can't pass the method directly like this,
```
execution_future_ = std::async(std::launch::async, my_new_work_method);
```
then you can use a simple lambda here to call the method
```
execution_future_ = std::async(std::launch::async, []() {my_new_work_method();} );
```

You can do the same sort of thing for `handle_cancel` and `handle_accepted`.